Add informer-based controller

pmorie opened the pull request:
Create a new informer-based controller and a new binary to run it, along with accompanying config structs and all the trimmings.

Depends on #340.

krancour commented:
Note that since #340 was merged, 7ae91683cf88ca8e79ae48e98d83b9c2129283d5 doesn't need to be included in this PR.

krancour commented:
Also wanted to capture here that @pmorie and I discussed offline (some time ago) and agreed the package name (`openservicebroker`) needs to change, as it currently hints that there's something OSB-specific about this controller, but really the controller doesn't know anything about OSB.

pmorie commented on cmd/controller-manager/app/options/options.go:
```diff
@@ -0,0 +1,51 @@
+import (
+	"github.com/kubernetes-incubator/service-catalog/pkg/apis/componentconfig"
+	"github.com/spf13/pflag"
+	k8scomponentconfig "k8s.io/kubernetes/pkg/apis/componentconfig"
```
oops, used the wrong package.

krancour commented on pkg/apis/componentconfig/types.go:
```diff
@@ -30,3 +30,33 @@ type ControllerConfiguration struct {
+type ControllerManagerConfiguration struct {
+	// Address is the IP address to serve on (set to 0.0.0.0 for all interfaces).
+	Address string
+	// Port is the port that the controller's http service runs on.
```
Nit: s/controller's/controller manager's/

krancour commented on cmd/controller-manager/app/controller.go:
```diff
@@ -0,0 +1,279 @@
+/*
```
Rename this file to `controller_manager.go`?

arschles commented on cmd/controller-manager/app/controller.go:
```diff
@@ -0,0 +1,279 @@
+		Long: `The service-catalog controller manager is a daemon that embeds
+the core control loops shipped with the service catalog.`,
+		Run: func(cmd *cobra.Command, args []string) {
+		},
```
this does nothing?

arschles commented on pkg/controller/client_builder.go:
```diff
@@ -0,0 +1,84 @@
+	"github.com/golang/glog"
+)
+
+// ClientBuilder allow syou to get clients and configs for controllers
```
typo nit: "allow syo" => "allows you"

arschles commented on pkg/controller/wip/controller.go:
```diff
@@ -0,0 +1,204 @@
+	bindingInformer      cache.SharedInformer
+}
+
+var _ Controller = &controller{}
```
This is unnecessary. The above `NewController` function will ensure that `controller` implements `Controller`

krancour commented:
@pmorie should there be some chart updates in here also so that the new controller-manager gets deployed instead of the old controller?

pmorie commented:
@krancour 

> should there be some chart updates in here also so that the new controller-manager gets deployed instead of the old controller?

Good question.  I would like to just add in the controller at this point and establish that it's working before updating things to use it.  Updating charts is something I'd like to coordinate with people about before starting on.

krancour commented:
@pmorie so everyone who wants to verify this works needs to go modify the chart themselves? It seems it would be more efficient to bundle it in here.

pmorie commented:
@krancour I'm going to be working on testing this tomorrow -- let's talk about chart updates then, if you have some time.

pmorie commented:
I've addressed the nits that have been left so far.  This is currently working correctly for me using the `catalog-wip` chart using the following helm command to install:

```
helm install --namespace=service-catalog --set version=latest,debug=1,nodePort=30000,imagePullPolicy=IfNotPresent,insecure=true ~/code/service-catalog/deploy/wip-catalog/ -n sc
```

I suggest that we merge this and being porting in business logic and address nits later in follow-ups.

arschles commented:
I am making the administrative decision to merge this now, and follow @pmorie's suggestion to address nits and business logic porting in follow-ups.
